Convenience changes, tuple region macro

antiguru opened the pull request:
This PR adds convenience implementations for `ColumnStack` and:
* Macro to generate the tuple regions
* More types
* destructured copy for tuples.

antiguru committed:
Destructured copy and convenience impls

Signed-off-by: Moritz Hoffmann <antiguru@gmail.com>

antiguru committed:
Macro to derive columnation for tuples

Signed-off-by: Moritz Hoffmann <antiguru@gmail.com>

antiguru committed:
Support additional types

Signed-off-by: Moritz Hoffmann <antiguru@gmail.com>

frankmcsherry commented on src/lib.rs:
```diff
@@ -214,6 +227,34 @@ impl<T: Columnation> ColumnStack<T> {
     }
 }
 
+impl<A: Columnation, B: Columnation> ColumnStack<(A, B)> {
+    /// Copies a destructured tuple `(A, B)` into this column stack.
+    ///
+    /// This serves situations where a tuple should be constructed from its constituents but not
+    /// not all elements are available as owned data.
+    ///
+    /// The element can be read by indexing
+    pub fn copy_destructured(&mut self, t1: &A, t2: &B) {
+        unsafe {
+            self.local.push(self.inner.copy_destructured(t1, t2));
+        }
+    }
+}
+
+impl<A: Columnation, B: Columnation, C: Columnation> ColumnStack<(A, B, C)> {
+    /// Copies a destructured tuple `(A, B, C)` into this column stack.
+    ///
+    /// This serves situations where a tuple should be constructed from its constituents but not
+    /// not all elements are available as owned data.
+    ///
+    /// The element can be read by indexing
+    pub fn copy_destructured(&mut self, r0: &A, r1: &B, r2: &C) {
+        unsafe {
+            self.local.push(self.inner.copy_destructured(r0, r1, r2));
+        }
+    }
+}
+
```
Nit, but was this an example of a thing that could be generalized with `paste` as well?

antiguru replied:
Converted to macro.

frankmcsherry commented on src/lib.rs:
```diff
@@ -237,6 +278,53 @@ impl<T: Columnation> Default for ColumnStack<T> {
+    }
+}
+
+impl<A: Columnation, B: Borrow<A>> std::iter::FromIterator<B> for ColumnStack<A> {
+    fn from_iter<T: IntoIterator<Item=B>>(iter: T) -> Self {
+        let iter = iter.into_iter();
+        let mut c = ColumnStack::<A>::with_capacity(iter.size_hint().0);
+        c.extend(iter);
+        c
+    }
+}
```
Nit, but elsewhere `T` is the `Columnation` type. Not always, now looking at it, but often.

antiguru replied:
Fixed, and removed the `Borrow` requirement. It turns out it's difficult to use this way.

frankmcsherry commented on src/lib.rs:
```diff
@@ -175,6 +177,17 @@ pub struct ColumnStack<T: Columnation> {
 }
 
 impl<T: Columnation> ColumnStack<T> {
+    /// Construct a [ColumnStack], reserving space for `capacity` elements
+    ///
+    /// Note that the associated region is not initialized to a specific capacity
+    /// because we can't generally know how much space would be required.
+    pub fn with_capacity(capacity: usize) -> Self {
+        Self {
+            local: Vec::with_capacity(capacity),
+            inner: T::InnerRegion::default(),
+        }
+    }
+
```
I think we discussed this, and IIRC one take was that regions could have a `fn with_capacity_for(&[T]) -> Self` or something analogous, maybe an iterator, which would enable a better `clone_from` implementation. If we concluded that, no need to fix here but it might be worth leaving a note (IIRC, this method mainly exists to be used by another PR elsewhere, I think?).

antiguru replied:
The function is private now, because it's useful internally but doesn't do what a caller would expect it to do. A `fn with_capacity_for(&[T]) -> Self` seems useful, but it would require support of the regions, which we currently don't have. I'll leave this for later.

frankmcsherry reviewed (approved):
These look good to me. A few comments that I wanted to sort out at some point, but needn't block merging!

antiguru committed:
ColumnStack: extend/from_iter without Borrow

It turns out to be rather difficult to have a generic version of extend and
from_iter for Borrow items, so use references directly.

Signed-off-by: Moritz Hoffmann <antiguru@gmail.com>

antiguru committed:
ColumnStack: Macro for copy_destructured

We also export the macro for use by similar implementations of ColumnStack.

Signed-off-by: Moritz Hoffmann <antiguru@gmail.com>

antiguru committed:
ColumnStack: privatize with_capacity

with_capacity is misleading in that it only allocates the local part, but
does not reserve space for the inner region. For this reason, it seems
better to not advertise the function to the outside.

Signed-off-by: Moritz Hoffmann <antiguru@gmail.com>

antiguru commented:
I addressed your comments and opened #7 to capture the `with_capacity_for` idea. Thanks a bunch!

frankmcsherry commented:
Thanks very much!
